### ISSUE TITLE
Custom type renderers are not reused, but mounted/unmounted excessively

### ISSUE THREAD

**Issue by vramdal** [sanity-io/block-content-to-react#6]
I see a lot of unmounting/re-mounting of my component that is used in a custom serializer. In my mind, the component should not be unmounted/remounted on every props update, but instead just updated with new props.

Given the following app (or see [sandbox](https://codesandbox.io/s/p3zlzvrk4q)):
```
import ReactDOM from 'react-dom';
import React, {Component} from "react";
import BlockContent from "@sanity/block-content-to-react";

class Sometype extends Component {
    
    componentDidMount() {
        console.log("Sometype.componentDidMount - this should only happen once", arguments);
    }

    componentWillReceiveProps(nextProps) {
        console.log("Sometype.componentWillReceiveProps - this should happen on every click, but never happens", arguments);
    }

    render() {
        return <p onClick={this.props.clickHandler}>Click me! {this.props.children}</p>
    }
}

const block = {
    "_key":"58cf8aa1fc74",
    "_type":"sometype",
    "something":{someProperty: "someValue"}
};

class App extends Component {
    constructor() {
        super();
        this.state = {clicks: 0};
        this.serializers = {
            types: {
                sometype: props => (
                    <Sometype clickHandler={() => this.setState({clicks: this.state.clicks + 1})}>
                        {props.node.something.someProperty + " " + this.state.clicks}
                    </Sometype>
                )
            }
        };
    }


    render() {
        return (
            <div className="App">
                <BlockContent key={'blockcontent-1'} blocks={[block]} serializers={this.serializers}/>
            </div>
        );
    }
}

ReactDOM.render(<App />, document.getElementById('root'));
```

The app uses `BlockContent` with a map of a single serializer for type `sometype`. The serializer renders a `Sometype` component.

In the browser, when clicking any content from `Sometype`, the state of the container component is updated, thus triggering a re-render.

At this point, I would expect `Sometype.componentWillUpdate` to be called, with the new updated props. Instead, the `Sometype` instance is unmounted, and a new instance is mounted, thus never calling `componentWillUpdate`. I'm not sure this violates any spec of a React component, but it is certainly uncommon behaviour.

Compare to [an app](https://codesandbox.io/s/kokxz571yo) _not_ using `BlockContent`, where `Sometype.componentWillUpdate` _is_ called on props update.

By modifying `block-content-to-hyperscript/lib/serializers.js` so that it returns an object instead of a function (thus hardcoding `React.createElement` instead of passing it as a parameter to the function), and modifying `BlockContent` to used the exported object instead of calling the function, the problem goes away, and `Sometype.componentWillUpdate` is called as expected. 

[This article](https://www.huy-nguyen.com/blog/2017-01-avoid-unnecessary-remounting-react) describes a somewhat similar problem, but I am unable to see how it can apply to this problem.

**Comment by vramdal** [comment]
The problem seems to be related to how [block-to-hyperscript's serializers.js](https://github.com/sanity-io/block-content-to-hyperscript/blob/master/src/serializers.js) returns a new serializer function on every occation. 

By [saving and reusing the serializer function](https://gist.github.com/vramdal/ac326a5bf5a4f2755dfe5e64125fceb1/revisions?diff=split), the problem can be avoided. But surely there must be a better way.

**Comment by vramdal** [comment]
`BlockContent` already gets `defaultSerializers` from calling `getSerializers(renderNode)`. If we let it also store a reference to `serializeSpan`,  we could pass those to functions to `blocksToNodes`, so that it would not have ot call `getSerializers` again:

`BlockContent.js`:
```diff
const renderNode = React.createElement
- const {defaultSerializers} = getSerializers(renderNode)
+ const {defaultSerializers, serializeSpan} = getSerializers(renderNode)

const SanityBlockContent = props => {
-  return blocksToNodes(renderNode, Object.assign({blocks: []}, props))
+  return blocksToNodes(renderNode, Object.assign({blocks: []}, props), defaultSerializers, serializeSpan)
}
```

`blocksToNodes.js`:
```diff
- function blocksToNodes(h, properties) {
+ function blocksToNodes(h, properties, defaultSerializers, serializeSpan) {
-   const {defaultSerializers, serializeSpan} = getSerializers(h)
```

WDYT? If this seems like a good way to go, I could try putting together a PR.

**Comment by rexxars** [comment]
Thanks for a thorough bug report and for looking into it so deeply!

I think your solution seems like a good starting point seeing as it doesn't change the public API - would love a PR.

Looking at the code now, I think perhaps we should try to refactor things a bit so that the serializers are not created on each iteration, it seems excessive and costly - but I suppose we can do this later.

**Comment by rexxars** [comment]
Thanks again for reporting and for the fix! Really appreciated! Fixed in v1.3.7

**Comment by ibrhm-mnf-fingent** [comment]
Is this issue still exist in block-content-to-react?
Custom serializer component unmounting and mounts again on rerendering.

Codesandbox example updating a simple state after 3 seconds to show custom serializer component unmounting and remount.
https://codesandbox.io/s/sad-tu-fjgty?file=/src/App.js
